From a8e6b417868dc4530d2669d6870c1817ea6af5e7 Mon Sep 17 00:00:00 2001
Message-ID: <a8e6b417868dc4530d2669d6870c1817ea6af5e7.1777630652.git.lorenzo@kernel.org>
In-Reply-To: <4f763b33f7e25eada0f919379e7dbe321f8a50a8.1777630652.git.lorenzo@kernel.org>
References: <4f763b33f7e25eada0f919379e7dbe321f8a50a8.1777630652.git.lorenzo@kernel.org>
From: Lorenzo Bianconi <lorenzo@kernel.org>
Date: Mon, 13 Apr 2026 17:38:51 +0200
Subject: [PATCH 9/9] net: airoha: Support multiple LAN/WAN interfaces for hw
 MAC address configuration

The EN7581 and AN7583 SoCs provide registers to configure hardware LAN/WAN
MAC addresses, used to determine whether received traffic is destined for
this host or should be forwarded to another device.
The SoC hardware design assumes all interfaces configured as LAN (or WAN)
share a common upper MAC address, which is programmed into the
REG_FE_{LAN,WAN}_MAC_H register. The lower bytes of 'local' addresses can
be expressed as a range via the REG_FE_MAC_LMIN and REG_FE_MAC_LMAX
registers.
Previously, only a single interface was considered when programming these
registers. Extend the logic to derive the correct minimum and maximum
values for REG_FE_MAC_LMIN/REG_FE_MAC_LMAX when two or more interfaces are
configured as LAN or WAN.

Signed-off-by: Lorenzo Bianconi <lorenzo@kernel.org>
---
 drivers/net/ethernet/airoha/airoha_eth.c | 73 ++++++++++++++++++++----
 1 file changed, 63 insertions(+), 10 deletions(-)

--- a/drivers/net/ethernet/airoha/airoha_eth.c
+++ b/drivers/net/ethernet/airoha/airoha_eth.c
@@ -79,20 +79,67 @@ static bool airhoa_is_phy_external(struc
 }
 #endif
 
-static void airoha_set_macaddr(struct airoha_gdm_dev *dev, const u8 *addr)
+static int airoha_set_macaddr(struct airoha_gdm_dev *dev, const u8 *addr)
 {
 	struct airoha_eth *eth = dev->eth;
-	u32 val, reg;
+	u8 ref_addr[ETH_ALEN] = {};
+	u32 reg, val, lmin, lmax;
+	int i;
+
+	lmin = (addr[3] << 16) | (addr[4] << 8) | addr[5];
+	lmax = lmin;
+
+	for (i = 0; i < ARRAY_SIZE(eth->ports); i++) {
+		struct airoha_gdm_port *port = eth->ports[i];
+		int j;
+
+		if (!port)
+			continue;
+
+		for (j = 0; j < ARRAY_SIZE(port->devs); j++) {
+			struct airoha_gdm_dev *iter_dev;
+			struct net_device *netdev;
+
+			iter_dev = port->devs[j];
+			if (!iter_dev || iter_dev == dev)
+				continue;
+
+			if (airoha_is_lan_gdm_dev(iter_dev) !=
+			    airoha_is_lan_gdm_dev(dev))
+				continue;
+
+			netdev = iter_dev->dev;
+			if (netdev->reg_state != NETREG_REGISTERED)
+				continue;
+
+			ether_addr_copy(ref_addr, netdev->dev_addr);
+			val = (netdev->dev_addr[3] << 16) |
+			      (netdev->dev_addr[4] << 8) | netdev->dev_addr[5];
+			if (val < lmin)
+				lmin = val;
+			if (val > lmax)
+				lmax = val;
+		}
+	}
+
+	if (!is_zero_ether_addr(ref_addr) && memcmp(ref_addr, addr, 3)) {
+		/* According to the HW design, hw mac address MS bits
+		 * must be the same for each net_device with the same
+		 * LAN/WAN configuration.
+		 */
+		return -EINVAL;
+	}
 
 	reg = airoha_is_lan_gdm_dev(dev) ? REG_FE_LAN_MAC_H : REG_FE_WAN_MAC_H;
 	val = (addr[0] << 16) | (addr[1] << 8) | addr[2];
 	airoha_fe_wr(eth, reg, val);
 
-	val = (addr[3] << 16) | (addr[4] << 8) | addr[5];
-	airoha_fe_wr(eth, REG_FE_MAC_LMIN(reg), val);
-	airoha_fe_wr(eth, REG_FE_MAC_LMAX(reg), val);
+	airoha_fe_wr(eth, REG_FE_MAC_LMIN(reg), lmin);
+	airoha_fe_wr(eth, REG_FE_MAC_LMAX(reg), lmax);
 
 	airoha_ppe_init_upd_mem(dev, addr);
+
+	return 0;
 }
 
 static void airoha_set_gdm_port_fwd_cfg(struct airoha_eth *eth, u32 addr,
@@ -1835,18 +1882,23 @@ static int airoha_dev_stop(struct net_de
 static int airoha_dev_set_macaddr(struct net_device *netdev, void *p)
 {
 	struct airoha_gdm_dev *dev = netdev_priv(netdev);
+	struct sockaddr *addr = p;
 	int err;
 
-	err = eth_mac_addr(netdev, p);
+	err = eth_prepare_mac_addr_change(netdev, p);
+	if (err)
+		return err;
+
+	err = airoha_set_macaddr(dev, addr->sa_data);
 	if (err)
 		return err;
 
-	airoha_set_macaddr(dev, netdev->dev_addr);
+	eth_commit_mac_addr_change(netdev, p);
 
 	return 0;
 }
 
-static int airoha_set_gdm2_loopback(struct airoha_gdm_dev *dev)
+static int airoha_enable_gdm2_loopback(struct airoha_gdm_dev *dev)
 {
 	struct airoha_gdm_port *port = dev->port;
 	struct airoha_eth *eth = dev->eth;
@@ -1989,33 +2041,38 @@ static int airoha_dev_init(struct net_de
 {
 	struct airoha_gdm_dev *dev = netdev_priv(netdev);
 	struct airoha_gdm_port *port = dev->port;
+	int err;
 
 	switch (port->id) {
-	case AIROHA_GDM2_IDX:
-		/* GDM2 is always used as wan */
-		dev->flags |= PRIV_FLAG_WAN;
-		break;
 	case AIROHA_GDM3_IDX:
 	case AIROHA_GDM4_IDX: {
 		struct airoha_eth *eth = dev->eth;
 
-		if (!eth->ports[1] && !airoha_get_wan_gdm_dev(eth)) {
-			int err;
-
-			err = airoha_set_gdm2_loopback(dev);
-			if (err)
-				return err;
-
-			dev->flags |= PRIV_FLAG_WAN;
-		}
-		break;
+		if (eth->ports[1] || airoha_get_wan_gdm_dev(eth))
+			break;
+		fallthrough;
 	}
+	case AIROHA_GDM2_IDX:
+		/* GDM2 is always used as wan */
+		dev->flags |= PRIV_FLAG_WAN;
+		break;
 	default:
 		break;
 	}
 
 	airoha_dev_set_qdma(dev);
-	airoha_set_macaddr(dev, netdev->dev_addr);
+	err = airoha_set_macaddr(dev, netdev->dev_addr);
+	if (err)
+		return err;
+
+	if (!airoha_is_lan_gdm_dev(dev) &&
+	    (port->id == AIROHA_GDM3_IDX || port->id == AIROHA_GDM4_IDX)) {
+		err = airoha_enable_gdm2_loopback(dev);
+		if (err) {
+			dev->flags &= ~PRIV_FLAG_WAN;
+			return err;
+		}
+	}
 
 	return 0;
 }
@@ -2301,6 +2358,7 @@ struct airoha_ethool_priv_flags {
 
 static int airoha_dev_set_wan_flag(struct net_device *netdev, u32 flags)
 {
+
 	struct airoha_gdm_dev *dev = netdev_priv(netdev);
 	struct airoha_gdm_port *port = dev->port;
 	struct airoha_eth *eth = dev->eth;
@@ -2321,22 +2379,26 @@ static int airoha_dev_set_wan_flag(struc
 
 		switch (port->id) {
 		case AIROHA_GDM2_IDX:
+			dev->flags |= PRIV_FLAG_WAN;
+			airoha_dev_set_qdma(dev);
 			break;
 		case AIROHA_GDM3_IDX:
 		case AIROHA_GDM4_IDX: {
 			int err;
 
-			err = airoha_set_gdm2_loopback(dev);
-			if (err)
+			dev->flags |= PRIV_FLAG_WAN;
+			airoha_dev_set_qdma(dev);
+
+			err = airoha_enable_gdm2_loopback(dev);
+			if (err) {
+				dev->flags &= ~PRIV_FLAG_WAN;
 				return err;
+			}
 			break;
 		}
 		default:
 			return -EOPNOTSUPP;
 		}
-
-		dev->flags |= PRIV_FLAG_WAN;
-		airoha_set_macaddr(dev, netdev->dev_addr);
 	} else {
 		switch (port->id) {
 		case AIROHA_GDM3_IDX:
@@ -2351,12 +2413,12 @@ static int airoha_dev_set_wan_flag(struc
 		default:
 			break;
 		}
+
 		dev->flags &= ~PRIV_FLAG_WAN;
+		airoha_dev_set_qdma(dev);
 	}
 
-	airoha_dev_set_qdma(dev);
-
-	return 0;
+	return airoha_set_macaddr(dev, netdev->dev_addr);;
 }
 
 static const struct airoha_ethool_priv_flags airoha_eth_priv_flags[] = {
